libs/live-wallet/src/walletsync/incrementalUpdates.ts

```diff
@@ -0,0 +1,89 @@
+import { log } from "@ledgerhq/logs";
+import walletsync, { DistantState, LocalState } from ".";
+import { UpdateEvent } from "../cloudsync/sdk";
+import { WalletSyncDataManagerResolutionContext } from "./types";
+import { WSState } from "../store";
+
+export function makeSaveNewUpdate<S>({
+  ctx,
+  getState,
+  latestDistantStateSelector,
+  localStateSelector,
+  saveUpdate,
+}: {
+  ctx: WalletSyncDataManagerResolutionContext;
+  getState: () => S;
+  latestDistantStateSelector: (state: S) => DistantState | null;
+  localStateSelector: (state: S) => LocalState;
+  saveUpdate: (
+    data: DistantState | null,
+    version: number,
+    newLocalState: LocalState | null,
+  ) => Promise<void>;
+}): (event: UpdateEvent<DistantState>) => Promise<void> {
+  return async (event: UpdateEvent<DistantState>) => {
+    log("walletsync", "saveNewUpdate", { event });
+    switch (event.type) {
+      case "new-data": {
+        // we resolve incoming distant state changes
+        const state = getState();
+        const latest = latestDistantStateSelector(state);
+        const local = localStateSelector(state);
+        const data = event.data;
+        const resolved = await walletsync.resolveIncrementalUpdate(ctx, local, latest, data);
+
+        if (resolved.hasChanges) {
+          const version = event.version;
+          const localState = localStateSelector(getState()); // fetch again latest state because it might have changed
+          const newLocalState = walletsync.applyUpdate(localState, resolved.update); // we resolve in sync the new local state to save
+          await saveUpdate(data, version, newLocalState);
+          log("walletsync", "resolved. changes applied.");
+        } else {
+          log("walletsync", "resolved. no changes to apply.");
+        }
+        break;
+      }
+      case "pushed-data": {
+        await saveUpdate(event.data, event.version, null);
+        break;
+      }
+      case "deleted-data": {
+        await saveUpdate(null, 0, null);
+        break;
+      }
+    }
+  };
+}
+
+export function makeLocalIncrementalUpdate<S>({
+  ctx,
+  getState,
+  latestWalletStateSelector,
+  localStateSelector,
+  saveUpdate,
+}: {
+  ctx: WalletSyncDataManagerResolutionContext;
+  getState: () => S;
+  latestWalletStateSelector: (state: S) => WSState;
+  localStateSelector: (state: S) => LocalState;
+  saveUpdate: (
+    data: DistantState | null,
+    version: number,
+    newLocalState: LocalState | null,
+  ) => Promise<void>;
+}): () => Promise<void> {
+  return async () => {
+    // we resolve possible local incremental update
+    const state = getState();
+    const { data, version } = latestWalletStateSelector(state);
+    const local = localStateSelector(state);
+    const resolved = await walletsync.resolveIncrementalUpdate(ctx, local, data, data);
+
+    if (resolved.hasChanges) {
+      const localState = localStateSelector(getState()); // fetch again latest state because it might have changed
```

Comment:
Is it not risky to rely on new values of the state? Because here we could have `resolved` that would be computed on a previous version of the state and applied in applyUpdate with a different version of the state.

Reply:
@KVNLS on the generic part (of that code you point out) everything is fine, as long at the module have an "incremental" implementation of an "update". so it all depends on how the module is implemented:
- for `accounts`, i think we are safe because we express an "update" as a really incremental thing, instead of saying "here is the new accounts" in a resolve, we say "here is the added, here is the deleted", and we apply on top of latest state these "actions".
- for `accountNames`, **in the current naive implementation** there is a risk of overriding the account names: if you modify the account names on both sides at the same time, one will win over the other, because the module is currently very naive: it replaces all the names it receives without taking care of the one modified. => i'll write a ticket on this one for future improvement. IMO it's not really a big issue because you rarely modify account names at the same time on both sides.

Reply:
Thanks for the explanation! Do we have unit tests for both use cases?

Reply:
*also, the answer i just gave is a generic comment on the current implementation. the update of this PR have not introduced or regress on this and it not actually related to the code added here: because the code here is exclusively going to be effective on `accounts` for the case of "pending non imported accounts" to handle, as for the accountNames, it is a noop because the module bail out the case where "latestState === incomingState" in resolveIncrementalUpdate

Reply:
@KVNLS yes, we have a lot of unit tests around modules. both "generic" (generatively test them on tons of data) and "specific" ones (cover specific usecases and expectations).